lib/cn/move.c

```diff
@@ -29,54 +29,55 @@ cn_move(
     bool                       src_del,
     struct cn_tree_node       *tgt_node)
 {
-    struct cn_tree *tree;
-    struct kvset_list_entry *src, *src_end, *tgt;
+    struct kvset_list_entry *src, *tgt;
     struct list_head *src_head, *tgt_head;
-    uint64_t *src_ksidv = NULL;
+    struct cn_tree *tree;
     uint64_t src_nodeid;
-    merr_t err;
 
     INVARIANT(w && src_node && tgt_node);
-    INVARIANT((src_cnt > 0 && src_list) || (src_cnt == 0 && !src_list));
+    assert((src_cnt > 0 && src_list) || (src_cnt == 0 && !src_list));
 
     tree = w->cw_tree;
     src_head = &src_node->tn_kvset_list;
     tgt_head = &tgt_node->tn_kvset_list;
 
     src = src_list;
-    src_end = NULL;
     if (src) {
-        src_ksidv = malloc(sizeof(*src_ksidv) * src_cnt);
-        if (!src_ksidv)
-            return merr(ENOMEM);
-
+        uint64_t src_ksidv[src_cnt];
+        merr_t err;
+        void *lock;
+
+        /* Must hold the tree lock while iterating over the kvset list
+         * (because if src_node is the cn root node it could change).
+         */
+        rmlock_rlock(&tree->ct_lock, &lock);
         for (uint32_t i = 0; i < src_cnt; i++) {
             assert(src);
+            assert(kvset_get_work(src->le_kvset) == w);
+
             src_ksidv[i] = kvset_get_id(src->le_kvset);
             src = list_next_entry_or_null(src, le_link, src_head);
         }
-        src_end = src;
+
+        assert(!src_del || !src);
+        rmlock_runlock(lock);
 
         err = cndb_record_kvsetv_move(cn_get_cndb(tree->cn), tree->cnid, src_node->tn_nodeid,
                                       tgt_node->tn_nodeid, src_cnt, src_ksidv);
-        if (err) {
-            free(src_ksidv);
+        if (err)
```

Comment:
@nabeelmmd Does that rework of the assert capture your original intent?

Reply:
yes, it does!